.travis.yml

```diff
@@ -8,60 +8,70 @@ dist: trusty
-      echo "Incoming pull request from https://github.com/$TRAVIS_REPO_SLUG/pull/$TRAVIS_PULL_REQUEST";
-      author=$(curl -u dummy4dummy:dummy2dummy -s "https://api.github.com/repos/$TRAVIS_REPO_SLUG/pulls/$TRAVIS_PULL_REQUEST" | jq -r ".user.login");
-      if [ $? -ne 0 ] ; then exit 1; fi;
-      echo "Pull request submitted by $author";
-      signed=$(curl -s https://www.lightbend.com/contribute/cla/scala/check/$author | jq -r ".signed");
-      if [ $? -ne 0 ] ; then exit 1; fi;
-      if [ "$signed" = "true" ] ; then
-        echo "CLA check for $author successful";
-      else
-        echo "CLA check for $author failed";
-        echo "Please sign the Scala CLA to contribute to Scala Native";
-        echo "Go to https://www.lightbend.com/contribute/cla/scala and then resubmit this pull request";
-        exit 1;
-      fi;
-    fi;
-
 script:
   - java -version
-  - if [ -n "$CLANG_FORMAT_PATH" ]; then scripts/clangfmt --test; fi
-  - scripts/scalafmt --test
-  - >
-    sbt \
-      "-no-colors" \
-      "-J-Xmx2G" \
-      "^^ $SBT_VERSION" \
-      rebuild \
-      "set scriptedBufferLog in sbtScalaNative := false" \
-      test-all
+  - if [ "$MODE" == "source-checks" ]; then
+      if [ -n "$CLANG_FORMAT_PATH" ]; then
+        scripts/clangfmt --test;
+        if [ $? -ne 0 ] ; then exit 1; fi;
+      fi;
+      scripts/scalafmt --test;
+      if [ $? -ne 0 ] ; then exit 1; fi;
+      if [ "$TRAVIS_PULL_REQUEST" != "false" ] ; then
+        echo "Incoming pull request from https://github.com/$TRAVIS_REPO_SLUG/pull/$TRAVIS_PULL_REQUEST";
+        author=$(curl -u dummy4dummy:dummy2dummy -s "https://api.github.com/repos/$TRAVIS_REPO_SLUG/pulls/$TRAVIS_PULL_REQUEST" | jq -r ".user.login");
+        if [ $? -ne 0 ] ; then exit 1; fi;
+        echo "Pull request submitted by $author";
+        signed=$(curl -s https://www.lightbend.com/contribute/cla/scala/check/$author | jq -r ".signed");
+        if [ $? -ne 0 ] ; then exit 1; fi;
+        if [ "$signed" = "true" ] ; then
+          echo "CLA check for $author successful";
+        else
+          echo "CLA check for $author failed";
+          echo "Please sign the Scala CLA to contribute to Scala Native";
+          echo "Go to https://www.lightbend.com/contribute/cla/scala and then resubmit this pull request";
+          exit 1;
+        fi;
+      fi;
+    fi
+  - if [ "$MODE" == "compile-tests" ]; then
+      sbt "-no-colors" "-J-Xmx2G" "^^ $SBT_VERSION" nscplugin/publishLocal sbtScalaNative/publishLocal test-compile;
+    fi
+  - if [ "$MODE" == "runtime-tests" ]; then
+      ./ci-docker/load-cached-images.sh
+      docker run --rm --privileged multiarch/qemu-user-static:register;
+      docker build --cache-from scala-native-testing:linux-$TARGET_DOCKER_PLATFORM -t scala-native-testing:linux-$TARGET_DOCKER_PLATFORM --build-arg TARGET_DOCKER_PLATFORM=$TARGET_DOCKER_PLATFORM --build-arg HOST_DEBIAN_PLATFORM=amd64 ci-docker;
+      ./ci-docker/run-test-travis.sh;
```

Comment:
The main reason I didn't go with that approach was to prevent duplication with definition of environment variables inside the docker-compose config but also referencing the same information when identifying which container to run in Travis. I'll look at setting up a script that automatically builds and runs a specific platform test though.

Reply:
As I see it you either duplicate it in `.travis.yml` or in the docker-compose.yml file. To me the whole point of using docker is to allow reproducing the build environment everywhere, not just in Travis. Anyway, we will only need docker images for each combination of `x86_64`/`x86_32` and LLVM version. The rest can be managed with environment variables.